frontend/src/actions/transfers/transfer.ts

```diff
@@ -0,0 +1,187 @@
+      ? new RequestMetadata(data.requestMetadata)
+      : undefined;
+    this._requestFillMetadata = data.requestFillMetadata;
+  }
+
+  static new(
+    amount: TokenAmount,
+    sourceChain: Chain,
+    sourceToken: Token,
+    targetChain: Chain,
+    targetToken: Token,
+    targetAccount: EthereumAddress,
+    validityPeriod: UInt256,
+    fees: UInt256,
+  ): Transfer {
+    return new this({
+      amount: amount.encode(),
+      sourceChain,
+      sourceToken,
+      targetChain,
+      targetToken,
+      targetAccount,
+      validityPeriod: validityPeriod.encode(),
+      fees: fees.encode(),
+    });
+  }
+
+  get requestMetadata(): RequestMetadata | undefined {
+    return this._requestMetadata;
+  }
+
+  get requestFillMetadata(): RequestFillMetadata | undefined {
+    return this._requestFillMetadata;
+  }
+
+  protected getStepMethods(
+    signer: JsonRpcSigner,
+    signerAddress: EthereumAddress,
+  ): Record<string, CallableFunction> {
+    // For backwards compatibility, never remove an entry, only add new ones.
+    return {
+      makeRequestTransaction: () => this.makeRequestTransaction(signer, signerAddress),
+      waitForRequestEvent: () => this.waitForRequestEvent(),
+      waitForFulfillment: () => this.waitForFulfillment(),
+    };
+  }
+
+  public async execute(signer: JsonRpcSigner, signerAddress: EthereumAddress): Promise<void> {
+    const methods = this.getStepMethods(signer, signerAddress);
+    return super.executeSteps(methods);
+  }
+
+  public encode(): TransferData {
+    return {
+      amount: this.amount.encode(),
+      sourceChain: this.sourceChain,
+      sourceToken: this.sourceToken,
+      targetChain: this.targetChain,
+      targetToken: this.targetToken,
+      targetAccount: this.targetAccount,
+      validityPeriod: this.validityPeriod.encode(),
+      fees: this.fees.encode(),
+      steps: this.steps.map((step) => step.encode()),
+      requestMetadata: this._requestMetadata?.encode(),
+      requestFillMetadata: this.requestFillMetadata,
+    };
+  }
+
+  protected async makeRequestTransaction(
+    signer: JsonRpcSigner,
+    signerAddress: EthereumAddress,
+  ): Promise<void> {
+    const transactionHash = await makeRequestTransaction(
+      signer,
+      this.amount.uint256,
+      this.targetChain.identifier,
+      this.sourceChain.requestManagerAddress,
+      this.sourceToken.address,
+      this.targetToken.address,
+      this.targetAccount,
+      this.validityPeriod,
+      this.fees,
+    );
+
+    this._requestMetadata = new RequestMetadata({
+      transactionHash,
+      requestAccount: signerAddress,
+    });
+  }
+
+  protected async waitForRequestEvent(): Promise<void> {
+    if (!this._requestMetadata?.transactionHash) {
+      throw new Error('Attempt to get request event before sending transaction!');
+    }
+
+    const provider = new JsonRpcProvider(this.sourceChain.rpcUrl);
+    const identifier = await getRequestIdentifier(
```

Comment:
Hmm. You caught me. The reason was that we had these steps before and I tried to match them by splitting the functions according to them. 
Are you sure the `createRequest` call on the contract does not wait for the mining? Because that is what I was assuming.
But I'm open to change this.

Reply:
Very good point. This only worked because the roll-up sequencer is so fast. 
We fixed it both together. Fixup pushed.